src/dune_rules/link_time_code_gen.ml

```diff
@@ -12,9 +12,8 @@ let generate_and_compile_module cctx ~precompiled_cmi ~name ~lib ~code ~requires
   let* module_ =
     let+ modules = Dir_contents.modules_of_lib sctx lib in
     let obj_name =
-      Option.map modules ~f:(fun modules ->
-          let mli_only = Modules.find modules name |> Option.value_exn in
-          Module.obj_name mli_only)
+      Option.bind modules ~f:(fun modules ->
+          Option.map (Modules.find modules name) ~f:Module.obj_name)
```

Comment:
How about we allow the caller to provide an `obj_name` and touching `modules` if it's provided.